Add Changes and Item Changes Endpoints

srichter opened the pull request:
This implements the [Changes endpoint](https://developers.themoviedb.org/3/changes) for Movie, TV, and Person

srichter committed:
Implement checkOptionDate and checkOptionDateRange in Tmdb & Tests

srichter committed:
Implement Change and Change Result, Implement itemChangeGenerator

srichter committed:
Add Tests for Change endpoint

srichter committed:
PHPCSFixer changes, Fix comments in various files

srichter committed:
Add Tests for Change Result

vfalies commented on src/VfacTmdb/Tmdb.php:
```diff
@@ -353,4 +353,49 @@ public function checkOptionSessionId(array $options, array &$return) : void
+     */
+    public function checkOptionDate(string $option) : bool
+    {
+        $check = preg_match('/^([0-9]{4})-([0-9]{2}-[0-9]{2})$/', $option);
```
I'm not fan of this date checking. All cases are not covered. For instance, with this gestion, the date "2020-02-31" is valid, but it is not :)
The regex check a structure not a validity.
I suggest a validation function like:
```php
function validateDate($date, $format = 'Y-m-d H:i:s')
{
    $d = DateTime::createFromFormat($format, $date);
    return $d && $d->format($format) == $date;
}
```

srichter replied:
I agree, something like this would be way better. I will adjust

vfalies commented:
Thanks for the pull request !

Note: Exceptions seems not tested 
![image](https://user-images.githubusercontent.com/2021078/87443505-e356d700-c5f5-11ea-8fab-22965ea72e2d.png)

srichter commented:
Thanks for looking over the PR. I couldn't get the API to throw an error outside of sending a malformed date, so I'm not sure how we'd test the exceptions.

I also plan to implement the endpoints to receive the changes for a specific item ID, so maybe I will leave this PR open until I implement those as well.

srichter committed:
Refactor date option validation

srichter committed:
Add Tests for Exceptions in Change class

srichter committed:
Fix getAdult() in Change Result

srichter committed:
Refactor Change class to better describe its purpose

( It retrieves changes for a type, not for a specific item)

srichter committed:
Add support for retrieving changes to an item

- Implement ItemChanges Abstract Class
- Implement ItemChangeResultsInterface Interface
- Implement ItemChange Results Class
- Implement MovieItemChanges, PeopleItemChanges, TVItemChanges, TVSeasonItemChanges, TVEpisodeItemChanges Results Classes

srichter commented:
I've added the methods to retrieve changes for specific items as discussed in #53. I will leave some comments on areas where I feel there are things to explain to give you more context for your review. After you are satisfied with the implementation I will add the change methods to each item class as well as add tests.

srichter commented on src/VfacTmdb/Abstracts/Items/ItemChanges.php:
```diff
@@ -0,0 +1,103 @@
+     * @param string $key
+     * @return \Generator|Results\ItemChange
+     */
+    public function getChangesByKey(string $key) : \Generator
```
I feel this is a good method to have to allow users to only get changes that they care about. They could filter the results in their application instead, but we also use this method in TVItemChanges

srichter commented on src/VfacTmdb/Items/TVItemChanges.php:
```diff
@@ -0,0 +1,79 @@
+    /**
+     * @return \AppendIterator|Results\ItemChange
+     */
+    public function getEpisodeChanges()
```
I feel that both this and `getSeasonChanges()` are good methods to offer due to the way the API handles TV changes. Each change for a season or episode are listed as a change on the main TV item itself, but to find the specific changes you need to make additional requests with the returned ids until you get to the season/episode level. These helper functions will allow someone to access all of the episode changes for a TV show without having to handle the additional requests themselves. The methods may be better suited on the TVShow Item class.

srichter commented on src/VfacTmdb/Results/ItemChange.php:
```diff
@@ -0,0 +1,198 @@
+     * @param  \stdClass  $result
+     * @return \stdClass
+     */
+    private function initResultObject(\stdClass $result) : \stdClass
```
This function attempts to normalize the API data as the API is very inconsistent with the results here:

- Sometimes it does not include the `original_value` key.
- Sometimes it does not include the `value` key.
- Sometimes the `value` key is an int/string, other times it is an array.

srichter commented on src/VfacTmdb/Abstracts/Items/ItemChanges.php:
```diff
@@ -0,0 +1,103 @@
+        if (!empty($this->data->changes)) {
+            foreach ($this->data->changes as $change) {
+                foreach ($change->items as $item) {
+                    $item->key = $change->key;
```
The API groups changes by `key`; the part of the item which changed. Sometimes there will be multiple changes returned per key. I felt it made more sense to return each of those changes as their own ItemChange rather than an ItemChange being able to represent one or multiple changes and the user having to determine which. So we loop through each "key change" and append the key to it before creating the ItemChange.

vfalies commented:
I'm ok with your propositions but I thing that to upgrade code quality, the coverage of unit tests must be completed.

srichter committed:
Implement getChangeKeys method on ItemChanges Item Abstract

srichter committed:
Clean up code/comments, Rename TVItemChanges -> TVShowItemChanges

srichter committed:
Implement getItemChanges and getChanges methods on Items

- Movie, People, TVEpisode, TVSeason, TVShow

srichter committed:
Add Tests

srichter committed:
Apply Scrutinizer Fixes

srichter committed:
Update tests
